README.md
Hello world! My name is Ryan Huffman, and I made this game as part of an honors project.  
The project itself is based on a tutorial originally created by SentDex.  
> You can find his tutorial at  
> https://pythonprogramming.net/pygame-python-3-part-1-intro/  

I highly recommend checking it out if you want to build something like this yourself!

For those looking to jump right in, we'll keep it simple.  
> You will need to download and install Python  
> https://www.python.org/downloads/release/python-392/  

> You'll also need to download and install pygame  
> https://www.pygame.org/wiki/GettingStarted

In the game itself, you'll control the race car.  
You can use either 'a' & 'd', or the left and right arrows to move.  
If you want something else, feel free to modify the code to your liking.  
You do you. :D

While you're enjoying your pleasant drive,  
probably doing something noble like getting groceries for you grams,  
random things appear trying to stop you!  

Of course, you won't let that stop you.  
You'll make it past as many of these things as you can (within 3 lives).

I hope you enjoy!!!

----------------

This project is near and dear to me as it's a continuation of the project that got me into coding.  
It's the project that helped me convince myself to get into college,  
so being able to return to it was really fun!

This was the first time I had incorporated git and GitHub in a project,  
so that was a fun learning experience.  
It took a bit to understand commit notes, my readme was more of a changelog, etc.  
But by the end of the project I was more comfortable with the process of managing multiple branches,  
Was more specific with my commit notes, and streamlined the push process.  
This is also my first time using .md, and I can tell I'll need some practice,  
but I'm getting there.

The project itself was a nice change of pace.  
Instead of something scripted, the rehashed programs with known goals and values,  
it was nice to have a project that was only limited by my innovation,  
where I could take it anywhere I wanted.  
This also meant I had to go about it more strategically.  

It was clear after playing with the original code that I needed a plan.  
So I went to lucid charts to design myself a UML plan  
> Lucid Charts https://lucid.co/

One of the goals I set out was to clean up the game loop.  
I wanted a game loop that was mostly there to check variables and execute the proper code,  
but not to be the actual code itself.  
I wanted it to handoff control to other functions and classes for that,  
this way it was easy to follow the flow of information.  

I also wanted to convert the thing, player, and variables into their own classes.  
I knew I wanted to expand the features beyond just functions and variables.  
I wanted to encapsulate them and make them more robust.  

Finally, I wanted to split the data into two scripts,  
one for the core game logic and display related functions,  
and one for the tokens and variables.  
I wanted to expand on my experience working with interconnected files.  

So I got to rebuilding.  
Getting the initial files set up wasn't hard,  
but it required me to set up a lot of it before I could debug anything.  
After I could display the car, things got better.  
Then it was time to implement the things.  

I realized that having the variables in their own class was a horrible idea.  
(Ok, I went back and forth on this for some time)  
The inspiration was from my time with Java and C++ code structure.  
But, I realize, Python isn't Java or C++, and I shouldn't treat it as such.  
So I embraced public EVERYTHING and went to work with the variables up top.  

Hit some errors and combined the two scripts into one.  
Horrible idea.  
Split them up and cleaned up my code.  

A lot of time was spent trying to tune the numbers so there was more of a curve.  
With the numbers growing linearly, the challenge was ramping up to fast.  
It felt less like a game, and more like an insane gauntlet.  
But, with a few tweaks, (over a long period of time) I found a rhythm that was tolerable.  

From this point on it was just trying to play with and implement features that were lacking.  
I left the crash feature out for a while,  
relying on a message in the console to let me know when I crashed.  
So much easier to debug that way.  
(Who knew a car crash could take you out of your groove. ¯\_(ツ)_/¯ )  

The lane dividers took some work.  
I originally tried to repurpose the thing class to do it,  
creating all kinds of exceptions.  
Not the best way.  
Ended up giving them their own class to make my life easier.  
Balancing them took some time.  
It wasn't until I took away the "magic numbers" and based them off of resolution,  
(for an unrelated upgrade) that they clicked into place.  

Then to enhance I found a cool music track, implemented the crash functions,  
set up some lives, made a high score feature,  designed a continue screen, and continued tweaking.  
Creating a file for high scores required returning to my old book and searching the web a little.  
It's been awhile since I took my class. It was a good refresher!

I'll continue to work on this over time, so if you see any improvements,  
feel free to shoot a comment to me at rhuffman@mail.pima.edu or open a ticket!

------------

Data found in ClassObjects.py is game logic  
Data found in ARaceyDriveV2.py relates to flow and display.

Have fun playing with values in ClassObjects  
display_height/display_width (Your road, your rules)  
start2/start3 (The scores needed for more things)  
starting_score (It's not cheating, it's looking ahead!)  
FPS (clock speed modifier)  
global_multi (thing and car speed multi)  
lives (more is better, right?!?!)  

changelog.md
Feb 15   2hr  
I uploaded the initial state of the program.  
Still getting used to github so changes are expected.  
I'll continue to utilize this repo throughout the project and provide updates in this readme.  
Currently, it's just the core files. It will need an environment that contains pygame in order to run.  
I can include the virtual environment that I use if you want, but it looks like that's bad practice here.  

I'll try to figure out how to create a requirements file at some point.  
Expect a bulk of the improvements to occur during spring break.

March 4th  2hr
Added README.md to project (not just GitHub)  
fixed typo  
added colors list  
Tried working with classes...  
    It's slow progress  
Spent some time staring at the code...  
    Inspiration isn't my friend right tonight.  

March 6th  2hr  
Worked on UML diagrams  
Stared a lot mindlessly  
Worked on the car and thing classes  
Decided to scrap the data class and found I could just have the data declared up top to do what I wanted.  
Learned that two spaces at the end of a line in .md creates a new line!  
Decided that I'm going to rewrite a new version of the script instead of trying to re-engineer the original.  

March 15th 3.5h  
Tried to split the classes and some data into its own file again.  
Finding it challenging.  
Got them separated!  
Working on getting things to show up on the map.  
That took a minute. Now working staggering entrance.  
Test drove. Got bored around 150. Need to tweak challenge.  
Increased global multi from 0.5 to 0.8.  
1) That made it much more engaging. 
2) It reminded me I haven't implemented crash physics. Need to get on that  

Increased thing size from 50 to 75, decent change.
Played with thing speed  
Started implementing lines.  
  Kind of a disaster so far.

March 18th 2hr
Bringing in things sooner.  
Looked into lane dividers a bit. Back burner for now.    
Looked into adding music.  
Found a track called "Energy" from 
> https://freemusicarchive.org/music/Scott_Holmes/media-music-mix/energy-1  

Changed thing 2 start from 10 to 5 and thing 3 start from 30 to 15  
Implemented 3 lives and crash system.  
___Need to implement a continue screen.  
Fixed line dividers by making their own class.  

Mar 22nt 3 hours  
Added more crash messages  
Randomized messages  
Added smart logic to messages  
Made a continue screen  
Added optional size control argument to message_display()  
Decreased default message_display() size  
Implemented a persistent high score system  
Deleted old "Thing.py" file  

Mar 26  2 hours
Renamed this document to changelog.md  
Created Readme that actually describes the program and offers instructions on how to play  
Added reduce_numbers to Thing class to manage speed and size after crash  
Updated Car upgrade method  
Balanced several values  
Cleaned up code

Mar 30 2 hr
Tweaks and balancing  
Shrunk resolution  
Worked on making the dividers adjust with screen size  
Drastically reduced volume. From .5 - .1  
added colors  
worked on thing y reset_pos method  
Updated Readme.md  
Restructured the entire file structure for presentation  
Fixed bug

ARaceyDrive/ARaceyDriveV2.py
import pygame
import time
import random
import ClassObjects as CO


# Pygame data
pygame.init()
game_display = pygame.display.set_mode((CO.display_width, CO.display_height))
pygame.display.set_caption("A Racey Drive")
# Thank you Scott Holmes Music.
#  Track found at https://freemusicarchive.org/music/Scott_Holmes/media-music-mix/energy-1
try:
    pygame.mixer.music.load("Resources/Energy.mp3")
    pygame.mixer.music.set_volume(0.1)  # REMEMBER THIS! Almost blasted my ears out with it at full.
    pygame.mixer.music.play()
except C:
    print("An exception occurred during music playback.th music. Sorry.")
clock = pygame.time.Clock()


def display_score():
    """Display relevant data in top left"""
    font = pygame.font.SysFont("georgia", 25)
    text1 = font.render("High Score: " + str(CO.high_score), True, CO.white)
    text2 = font.render("Dodged:     " + str(CO.score), True, CO.white)
    text3 = font.render("Lives:      " + str(CO.lives), True, CO.white)
    game_display.blit(text1, (0, 0))
    game_display.blit(text2, (0, 20))
    game_display.blit(text3, (0, 40))


def message_display(text, size=75):
    """Used to display messages in the middle of the screen"""
    large_text = pygame.font.SysFont("georgia", size)
    text_surface = large_text.render(text, True, CO.black)
    text_rect = text_surface.get_rect()
    text_rect.center = ((CO.display_width/2), (CO.display_height/2))
    game_display.blit(text_surface, text_rect)

    pygame.display.update()


def crash(thing_list):
    # Base messages
    base1 = "You Can Do It"
    base2 = "Keep Trying"
    base = [base1, base2]

    # If almost high
    almost1 = "Almost There"
    almost2 = "You Can Do It"
    almost3 = "So Close"
    almost = [almost1, almost2, almost3]

    # If current high score
    hs1 = "Check You Out"
    hs2 = "Keep Going!"
    hs = [hs1, hs2]

    # Decides what message block to use based on distance from score
    messages = []
    if CO.score == CO.high_score:
        messages = hs
    elif CO.score > CO.high_score - 10:
        messages = almost
    else:
        messages = base

    message_display(random.choice(messages))

    # reduces speed and size of things
    for thing in thing_list:
        thing.reset_pos()
        thing.reduce_numbers()
    # resets divider speed to you can feel the velocity
    CO.Divider.speed = int(CO.display_height / 100)
    # cuts score in half
    CO.score = int(CO.score / 2)
    # reduces lives
    CO.lives -= 1
    # pauses so you can read the message
    # but not for long
    # you want to get people back in the action
    time.sleep(.5)


def continue_screen():
    to_continue = True
    keep_playing = True

    # display Encouragement
    # changes based on distance from high score
    game_display.fill(CO.lt_gray)
    if CO.score == CO.high_score:
        message_display("Try Again")
    elif CO.score > CO.high_score - 10:
        message_display("So Close")
    elif CO.score > CO.high_score - 25:
        message_display("Keep Pushing")
    else:
        message_display("Don't Give Up")
    pygame.display.update()
    time.sleep(.75)

    # display score
    game_display.fill(CO.lt_gray)
    message_display("Score: " + str(CO.score))
    pygame.display.update()
    time.sleep(.75)

    # display high score
    game_display.fill(CO.lt_gray)
    message_display("High Score: " + str(CO.high_score), 70)
    time.sleep(.75)

    while to_continue:
        # prompt for continue
        game_display.fill(CO.lt_gray)
        message_display("Continue? (y/n)", 65)

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()

            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_y or event.key == pygame.K_RETURN:
                    keep_playing = True
                    to_continue = False
                if event.key == pygame.K_n or event.key == pygame.K_ESCAPE:
                    keep_playing = False
                    to_continue = False

    return keep_playing


def game_loop():
    # sets base RUN variables
    keep_playing = True
    CO.update_local_high_score_from_file()

    while keep_playing:  # Outer loop for continue screen and to set starting conditions
        # sets base GAME variables
        CO.score = CO.starting_score
        CO.lives = CO.starting_lives

        # creates starter pieces
        player = CO.Car()
        thing_list = [CO.Thing(CO.greens, 1)]

        # lane dividers
        line_list = []
        for i in range(11):
            line_list.append(CO.Divider(CO.display_width*0.2, i*int(CO.display_height/10.5)))
            line_list.append(CO.Divider(CO.display_width*0.4, i*int(CO.display_height/10.5)))
            line_list.append(CO.Divider(CO.display_width*0.6, i*int(CO.display_height/10.5)))
            line_list.append(CO.Divider(CO.display_width*0.8, i*int(CO.display_height/10.5)))

        while CO.lives > 0:  # Inner loop for game
            # Background logic
            # display background based on score
            stage = int(CO.score / 50)
            if stage == 0:
                game_display.fill(CO.dk_green)
            elif stage == 1:
                game_display.fill(CO.green)
            elif stage == 2:
                game_display.fill(CO.dk_blue)
            if stage == 3:
                game_display.fill(CO.blue)
            elif stage == 4:
                game_display.fill(CO.dk_red)
            elif stage > 4:
                game_display.fill(CO.red)

            # draws track
            pygame.draw.rect(game_display, CO.dk_gray, [10, 0, CO.display_width - 20, CO.display_height])

            # displays white dividers
            for line in line_list:
                line.update()
                pygame.draw.rect(game_display, line.color, [int(line.x), int(line.y), line.width, line.height])

            # Thing logic
            # move things
            for thing in thing_list:
                # updates pos and checks for crash
                crashed = thing.update(player)
                if crashed:
                    # Adjusts score and sizes
                    crash(thing_list)
                else:
                    # draws thing in new pos
                    pygame.draw.rect(game_display, thing.color, [int(thing.x), int(thing.y), thing.width, thing.height])

            # adds new things at certain thresholds
            if CO.score >= CO.start2 and len(thing_list) == 1:
                thing_list.append(CO.Thing(CO.blues, 2))
            if CO.score >= CO.start3 and len(thing_list) == 2:
                thing_list.append(CO.Thing(CO.reds, 3))

            # display score (it's here so it remains on top of the things)
            display_score()

            # Player logic
            # check for speed increase
            player.upgrade()
            # update player pos
            player.update()
            game_display.blit(player.image, (player.x, player.y))

            # update screen
            pygame.display.update()
            clock.tick(CO.FPS)

        # once you lose all lives you'll reach the continue screen
        keep_playing = continue_screen()


if __name__ == "__main__":
    game_loop()

pygame.quit()
quit()


Original Code/ARaceyDrive.py
# The original code from https://pythonprogramming.net/adding-score-pygame-video-game/
# Car file and dimensions updated.

import pygame
import time
import random

pygame.init()

display_width = 800
display_height = 600

black = (0, 0, 0)
white = (255, 255, 255)
red = (255, 0, 0)

block_color = (53, 115, 255)

car_width = 29

gameDisplay = pygame.display.set_mode((display_width, display_height))
pygame.display.set_caption('A bit Racey')
clock = pygame.time.Clock()

carImg = pygame.image.load('../ARaceyDrive/Resources/aRaceyCar.png')


def things_dodged(count):
    font = pygame.font.SysFont(None, 25)
    text = font.render("Dodged: " + str(count), True, black)
    gameDisplay.blit(text, (0, 0))


def things(thingx, thingy, thingw, thingh, color):
    pygame.draw.rect(gameDisplay, color, [thingx, thingy, thingw, thingh])


def car(x, y):
    gameDisplay.blit(carImg, (x, y))


def text_objects(text, font):
    textSurface = font.render(text, True, black)
    return textSurface, textSurface.get_rect()


def message_display(text):
    largeText = pygame.font.Font('freesansbold.ttf', 115)
    TextSurf, TextRect = text_objects(text, largeText)
    TextRect.center = ((display_width / 2), (display_height / 2))
    gameDisplay.blit(TextSurf, TextRect)

    pygame.display.update()

    time.sleep(2)

    game_loop()


def crash():
    message_display('You Crashed')


def game_loop():
    x = (display_width * 0.45)
    y = (display_height * 0.8)

    x_change = 0

    thing_startx = random.randrange(0, display_width)
    thing_starty = -600
    thing_speed = 4
    thing_width = 100
    thing_height = 100

    thingCount = 1

    dodged = 0

    gameExit = False

    while not gameExit:

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()

            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_LEFT:
                    x_change = -5
                if event.key == pygame.K_RIGHT:
                    x_change = 5

            if event.type == pygame.KEYUP:
                if event.key == pygame.K_LEFT or event.key == pygame.K_RIGHT:
                    x_change = 0

        x += x_change
        gameDisplay.fill(white)

        # things(thingx, thingy, thingw, thingh, color)
        things(thing_startx, thing_starty, thing_width, thing_height, block_color)

        thing_starty += thing_speed
        car(x, y)
        things_dodged(dodged)

        if x > display_width - car_width or x < 0:
            crash()

        if thing_starty > display_height:
            thing_starty = 0 - thing_height
            thing_startx = random.randrange(0, display_width)
            dodged += 1
            thing_speed += 1
            thing_width += (dodged * 1.2)

        if y < thing_starty + thing_height:
            print('y crossover')

            if x > thing_startx and x < thing_startx + thing_width or x + car_width > thing_startx and x + car_width < thing_startx + thing_width:
                print('x crossover')
                crash()

        pygame.display.update()
        clock.tick(60)


game_loop()
pygame.quit()
quit()

ARaceyDrive/ClassObjects.py
"""
This file will hold the object data for the car, things, and data to get passed.
Not at all ready to move these out of the main file yet...
"""

import pygame
import random

# static Data
# some of these are fun to play with
high_score_file = "high_score.txt"
display_height = 800
display_width = 500
start2 = 5  # point at which 2nt thing arrives
start3 = 15  # point at which 3rd thing arrives
starting_score = 0
starting_lives = 3
car_speed_offset = 5  # (score/car_speed_multi) + car_speed_offset
car_speed_multi = 9  # (score/car_speed_multi) + car_speed_offset
thing_speed_offset = 10  # Thing start
FPS = 60
global_multi = 0.69  # affects top speed and growth curve

# dynamic Data
score = 0
high_score = 0
lives = 0

# colors
white = (255, 255, 255)
black = (0, 0, 0)
dk_gray = (60, 60, 60)
gray = (130, 130, 130)
lt_gray = (200, 200, 200)
lt_red = (255, 100, 100)
red = (255, 0, 0)
dk_red = (130, 0, 0)
reds = [red, dk_red, lt_red]
lt_green = (100, 255, 100)
green = (0, 255, 0)
dk_green = (0, 130, 0)
greens = [green, dk_green, lt_green]
lt_blue = (100, 100, 255)
blue = (0, 0, 255)
dk_blue = (0, 0, 130)
blues = [blue, dk_blue, lt_blue]


def get_high_score():
    """Get high score from file"""
    try:
        with open(high_score_file, 'r') as hs_txt:
            return int(hs_txt.read())
    except FileNotFoundError:
        return 0


def update_local_high_score_from_file():
    """Updates high score based on get_high_score"""
    global high_score
    current_hs = get_high_score()
    if current_hs is None or current_hs == "":
        pass
    elif int(current_hs) > high_score:
        high_score = int(current_hs)


def update_high_score_file():
    """Updates the high score file"""
    if high_score >= get_high_score():
        with open(high_score_file, 'w') as hs_txt:
            hs_txt.write(str(high_score))


def update_score():
    """Update score"""
    global score
    global high_score
    score += 1

    # update high score
    if score > high_score:
        high_score = score
        update_high_score_file()


class Car:
    """the car object the player controls"""
    def __init__(self):
        """Creates the car's start"""
        self.x = display_width*0.45
        self.y = display_height*0.8
        self.width = 29  # based on png, a little smaller
        self.height = 40  # based on png, a little smaller
        self.speed = car_speed_offset
        self.image = pygame.image.load("Resources/aRaceyCar.png")
        self.left = False
        self.right = False

    # get controls to move player
    # if you forget the keyup you get a fun constant mover. Enhancement?

    def get_controls(self):
        """Get the controls from user and update the proper flags"""
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                quit()

            # Idea for alteration, single button mode.
            # 'button' will toggle direction. Horizontal movement remains active

            # Another idea, adding up and down

            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_LEFT or event.key == pygame.K_a:
                    self.left = True
                    self.right = False
                if event.key == pygame.K_RIGHT or event.key == pygame.K_d:
                    self.left = False
                    self.right = True

            if event.type == pygame.KEYUP:
                if event.key == pygame.K_LEFT or event.key == pygame.K_a:
                    self.left = False
                if event.key == pygame.K_RIGHT or event.key == pygame.K_d:
                    self.right = False

    # command to update speed
    def upgrade(self):
        """Updates speed to be an offset of the score"""
        self.speed = (score / car_speed_offset) + car_speed_offset

    # update player position
    def update(self):
        """Moves car based on controls"""
        self.get_controls()
        # print("Car      ", int(self.speed), " ", int(pow(self.speed, global_multi)))       # diagnose feature
        if self.left:
            self.x -= int(pow(self.speed, global_multi))
        if self.right:
            self.x += int(pow(self.speed, global_multi))

        if self.x < 0:
            self.x = 0
        if self.x > display_width - self.width:
            self.x = display_width - self.width


class Divider:
    """Lane divider logic"""
    width = int(display_width / 60)  # Need to exp with these variables
    height = int(display_height / 40)
    color = white
    speed = int(display_height / 100)

    def __init__(self, x, y):
        """Set's x and y of divider"""
        self.x = x
        self.y = y

    def reset_pos(self):
        """Returns to top of screen"""
        self.y = -1 - self.height
        # print(Divider.speed)              # diagnosis feature
        if Divider.speed < 40:
            Divider.speed += .025

    def update(self):
        """Moves divider"""
        self.y += int(self.speed)

        if self.y > display_height + 5:
            self.reset_pos()


class Thing:
    """The things that tries to "block" you
    One will be designed to grow wider
    One will grow taller
    And the last will grow both ways"""

    def __init__(self, colors=None, t_id=0, width=75, height=75):
        """Set color list and id."""
        if colors is None:
            colors = greens
        self.colors = colors
        self.id = t_id
        self.width = width     # Need to exp with these variables
        self.height = height
        self.speed = int(thing_speed_offset + self.id + score / 3)  # adds extra speed if someone starts ahead
        self.color = random.choice(self.colors)
        self.x = random.randrange(int(self.width * .25), int(display_width - self.width * .75))
        self.y = - 10 - self.height

        if self.id == 1 or self.id == 3:  # adds extra thicccness if someone starts at a higher score.
            self.width += score / 3
        if self.id == 2 or self.id == 3:
            self.height += score / 3

    def reset_pos(self):
        """Resets pos to top of screen, changes color, and updates speed"""
        self.y = int(display_height / 4 * -1) - self.height
        self.x = random.randrange(int(self.width*.25), int(display_width - self.width * .75))
        self.color = random.choice(self.colors)
        self.speed += self.id

    def reduce_numbers(self):
        """Reduces size/speed during a crash event"""
        if self.id == 1:
            self.width = (self.width + 75) / 2
        elif self.id == 2:
            self.height = (self.height + 75) / 2
        elif self.id == 3:
            self.width = (self.width + 75) / 2
            self.height = (self.height + 75) / 2

        self.speed = int((self.speed + thing_speed_offset + self.id) / 2)

    def grow(self):
        """Causes the increase of height or width based on ID"""
        if self.id == 1:
            self.width += 1
        elif self.id == 2:
            self.height += 1
        elif self.id == 3:
            self.width += 1
            self.height += 1

    def update(self, car):
        """Updates thing position, checks if below bottom of screen or overlaps with car position"""
        is_crashed = False
        # print("Thing", self.id, " ", int(self.speed), " ", int(pow(self.speed, global_multi)))  # Diagnosis
        self.y += int(pow(self.speed, global_multi))

        # checks if below screen. resets pos, increases size, increases score if it is
        if self.y > display_height + 5:
            self.reset_pos()
            if self.id < 10:
                self.grow()
                update_score()

        # checks if crash has occurred
        if self.id < 10:
            if (self.y + 5) < car.y < (self.y + self.height - 5):
                if ((self.x + 5) < car.x < (self.x + self.width - 5)) \
                        or ((self.x + 5) < car.x + car.width < (self.x + self.width - 5)):
                    is_crashed = True

        # ends with result of crash
        return is_crashed

